### PR TITLE
1090 new patient modal fields are not consistent with the patient registration form 

### LINKED ISSUE #867: Customize items in generic modals
For example: Gender vs sex in patient creation dialog.

Image

Make this customizable

Idea: use JSONForms for it using UI schemas stored in form_schema table. Problem here is that currently only one schema is associated with a single document type.

This could be solved by removing the schema id from the `document_registry` and creating a `document_schema_join` table.
This table would have another context column like: `"DEFAULT" | "GENERIC_PATIENT_CREATION"`.

### LINKED ISSUE #1090: New Patient modal fields are not consistent with the Patient Registration form
### Dispensary > Patients > New Patient > New Patient Modal: 
- [x] there is no validation for the National Health ID (should be 10 characters, numerical only): 
Screen Shot 2023-01-25 at 2 11 20 PM

- [x] no validation for DOB:
Screen Shot 2023-01-25 at 2 12 29 PM

- [x] Gender dropdown of the modal should be the same as the one in the patient registration form
1. Patient Registration Form
Screen Shot 2023-01-25 at 2 14 15 PM

2. New Patient Modal
Screen Shot 2023-01-25 at 2 14 07 PM

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #867 & #1090

# 👩🏻‍💻 What does this PR do? 
- Add patient creation UI json schema
- Updates form schema endpoint to allow for query by filter
- Add form schema by type hook
- Update patient creation form to use the UI schema

# 🧪 How has/should this change been tested? 
Try to create a patient and everything should still work as it should 🤔

### CODE REVIEW

File: server/repository/src/db_diesel/form_schema_row.rs [hunk 1]
```diff
@@ -93,15 +95,18 @@ impl<'a> FormSchemaRowRepository<'a> {
     }
 
     #[cfg(not(feature = "postgres"))]
```
Comment (mark-prins) [inline server/repository/src/db_diesel/form_schema_row.rs:97]: line 86 also needs updating - currently won't compile when postgres feature is enabled

```
    pub fn upsert_one(&self, schema: &FormSchemaJson) -> Result<(), RepositoryError> {
```

File: server/service/src/sync/init_programs_data.rs [hunk 2]
```diff
@@ -1131,7 +1143,7 @@ pub fn init_program_data(
             name: "List of Appointments".to_string(),
             r#type: repository::ReportType::OmSupply,
             template: ENCOUNTERS_REPORT.to_string(),
-            context: ReportContext::Dispensary,
+            context: ReportContext::Patient,
```
Comment (mark-prins) [inline server/service/src/sync/init_programs_data.rs:1146]: why the change? 
There are now no results in the dispensary report list

**Review by mark-prins — CHANGES_REQUESTED** [review]
Overall.. awesome! nice and comprehensive and I like the approach.

Questions.. 
It is possible to select a date of birth which is in the future - but then the main patient form doesn't like that. Shouldn't the validation be the same on both?

The report context changes?

File: server/service/src/sync/init_programs_data.rs (hunk 2 above)
Comment (roxy-dao) [reply server/service/src/sync/init_programs_data.rs]: Not sure how these changes magically appeared 👀 

[641ed57](https://github.com/openmsupply/open-msupply/pull/1262/commits/641ed579eb72df6500624a298763daeadffea166)

File: client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx [hunk 3]
```diff
@@ -1,117 +1,57 @@
...
-
-  const genderOptions = useMemo(
-    () =>
-      Object.keys(Gender).map(key => ({
-        label: key,
-        value: key,
-      })),
-    []
+  const { data: patientCreationUI, isError, isLoading } = useFormSchema.document.byType(
+    'PatientCreationJSONForm'
```
Comment (clemens-msupply) [inline client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx:23]: Think it would be nice to build in a fallback, i.e. if the PatientCreationJSONForm is not configured and the returned data is undefined than use a very basic hardcoded default schema...

For the type I would call it `PatientCreationJSONForms` because the its JSONForms and not JSONForm...

**Review by clemens-msupply — COMMENTED** [review]
There is one bug I noticed: The last entered field is not taken over to the full form! no idea how this happens... For example, enter "firstName" and then "lastName" and lastName will be blank; same for other fields...

**Comment by roxy-dao** [comment]
> 641ed57

already added 😆

**Comment by clemens-msupply** [comment]
> There is one bug I noticed: The last entered field is not taken over to the full form! no idea how this happens... For example, enter "firstName" and then "lastName" and lastName will be blank; same for other fields...

Some update: its something with the debounce, for some reason the latest key stroke doesn't trigger an update, so maybe another bug will have another look...

File: client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx (hunk 3 above)
Comment (roxy-dao) [reply client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx:23]: You mean a default form schema, or just displaying the component that was there before if the creation form doesn't exist? 🤔

File: client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx [hunk 4]
```diff
@@ -1,117 +1,57 @@
...
-      Object.keys(Gender).map(key => ({
-        label: key,
-        value: key,
-      })),
-    []
+  const { data: patientCreationUI, isError, isLoading } = useFormSchema.document.byType(
+    'PatientCreationJSONForms'
   );
-  const dateFormatter = useFormatDateTime().customDate;
+  const [data, setData] = useState<Patient | undefined>();
```
Comment (clemens-msupply) [inline client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx:25]: Found the update bug: You call updatePatient with `data` and not with `newData`. `data` will only be updated in the next render cycle...

Think you don't need the useState at all do you?

File: client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx (hunk 3 above)
Comment (clemens-msupply) [reply client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx]: You could just have a hard coded json file for the default schemas in the UI code, think this would be the simplest solution(?)

File: client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx (hunk 4 above)
Comment (roxy-dao) [reply client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx:25]: I put it in a state so that the data can be passed through to the JsonForm component 🤔 

```suggestion
      const patientData = newData as Patient;
      updatePatient({
        code: patientData?.code,
        code2: patientData?.code2,
        firstName: patientData?.firstName,
        lastName: patientData?.lastName,
        dateOfBirth: patientData?.dateOfBirth,
        gender: patientData?.gender,
      });
```
^ having newData as Patient in a variable solves this too...

Comment (clemens-msupply) [reply client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx:25]: oh yes, you could use the patient from the usePatientCreateStore though...  but there might be some old data in this store(?) so having the useState is fine!

File: client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx (hunk 3 above)
Comment (mark-prins) [reply client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx]: yes, I did get this initially. shouldn't be a problem for normal usage though.

Screenshot 2023-03-02 at 3 57 23 PM

**Review by mark-prins — APPROVED** [review]

File: client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx (hunk 4 above)
Comment (roxy-dao) [reply client/packages/system/src/Patient/CreatePatientModal/PatientFormTab.tsx]: [ed60249](https://github.com/openmsupply/open-msupply/pull/1262/commits/ed60249ab0f093d3d723a22e5b17e688682ff032) 

🤔 I put a default patient creation in the FE...

**Review by clemens-msupply — APPROVED** [review]
Great work thank you!
